### PR TITLE
Optimize imports using global flag variable to ensure videojs and nanogallery are loaded once only

### PR DESCRIPTION
**Pull request by ibrahimjaved12**
# What are the relevant tickets?
https://github.com/mitodl/hq/issues/1749

# Description (What does it do?)
When we replaced `window.onload` with `window.addEventListener` to allow multiple listeners for `load`, we faced a new problem that a single kind of imports (eg. videojs imports) are called multiple times if there are multiple videos in a page. Not only that was not optimized, but this also caused some problems with the typescript functionality of download button popup (opening and closing) on videojs (online) videos.

This PR fixes this problem by using a global flag variable to keep track of videojs/nanogallery2 dynamic imports and to ensure they execute only once.

# How can this be tested?
1. Make sure the videos play + the video download button / related resources tab etc works as expected.

The functionality of the download button is as:
Click on video's download button, a popup should appear. Click on it again, it should close. Or, click outside the area, and it should close. Or click on another video's download button within the same page, the first video's popup should close and the other one's should open. (If it has Download Video or Download Transcript present)

The edge cases that caused problems:
1. If there were multiple dynamic imports present in a single page using `window.onload` only the last one would execute, hence we switched to using `AddEventHandler`. The original approach caused issues in the course where we had both image gallery+video present in a single page. (16.885j-fall-2005 --> related resources)
2. With the AddEventHandler, we had problem in multiple videos in a single page, because multiple import calls were made. (18.02sc-fall-2010 -- [57th session](https://live-qa.ocw.mit.edu/courses/18-02sc-multivariable-calculus-fall-2010/pages/3.-double-integrals-and-line-integrals-in-the-plane/part-b-vector-fields-and-line-integrals/session-57-work-and-line-integrals/)) Here we had problem we download button popup not working as expected, which is the reason for this PR.

And then of course we will want to handle these and every case for offline theme as well. 
**The steps to test this would be same as this: https://github.com/mitodl/ocw-hugo-themes/pull/1185 but a more thorough test with regards to videojs/nanogallery2 functionalities such as combinations of such different edge cases.**

Now as I understand, for offline theme, we do not have the video download button working as part of functionality because it's offline theme. And if we have other video tabs, like related resources, or "view video page", we would show the download icon on the right but it still shouldn't work. And if we don't have those additional tabs, then we wouldn't have the tab with download button icon at all. Please confirm this @ChristopherChudzicki.

### CODE REVIEW

File: course-offline/assets/course-offline.ts [hunk 1]
```diff
@@ -18,6 +17,7 @@ import "nanogallery2/src/jquery.nanogallery2.core.js"
 import "nanogallery2/src/css/nanogallery2.css"
 import "videojs-youtube"
 import videojs from "video.js"
+import _ from "lodash"
```
Comment (ChristopherChudzicki) [inline course-offline/assets/course-offline.ts:20]: @ibrahimjaved12 My apologies for suggesting a lodash approach. I had been under the impression we were already using lodash in the course bundle, but I guess we aren't. Given that, I think it's better to not use `_.once` and just avoid the double initialization with a simple flag.

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]

File: course-offline/assets/course-offline.ts [hunk 2]
```diff
@@ -27,6 +27,8 @@ export interface OCWWindow extends Window {
 
 declare let window: OCWWindow
 
+const initVideoJS = _.once(() => (window.videojs = videojs))
+
```
Comment (ChristopherChudzicki) [inline course-offline/assets/course-offline.ts:31]: Is there a reason to make this change? I don't think the `window.videojs = videojs` assignment was problematic.

File: course-v2/assets/course-v2.ts [hunk 3]
```diff
@@ -26,13 +34,14 @@ $(function() {
   initCourseDrawersClosingViaSwiping()
 })
 
-// @ts-expect-error for window.initVideoJS()
+window.videoJSLoaded = false
+window.nanogallery2Loaded = false
```
Comment (ChristopherChudzicki) [inline course-v2/assets/course-v2.ts:38]: I don't think there's any reason to put these on the global `window` object. We could scope them:

```ts
let videoJSLoaded = false
window.initVideoJS = () => {
  if (videoJSLoaded) return
  initVideoDownloadPopup()
  import("./videojs-imports").then(module => {
    module.initVideoJS()
  })
}
```

Or we could create a helper version of `once` without lodash:
```ts
const once = (callback: () => void) => {
  let hasRun = false
  if (!hasRun) {
    callback()
  }
}
```

Either way, I think it's worth putting fewer things on the global object.

**Review by ChristopherChudzicki — CHANGES_REQUESTED** [review]

**Review by ChristopherChudzicki — APPROVED** [review]

File: course-v2/assets/course-v2.ts [hunk 4]
```diff
@@ -26,13 +32,20 @@ $(function() {
...
     module.initVideoJS()
   })
+  videoJSLoaded = true
 }
 
-// @ts-expect-error for window.initNanogallery2()
-window.initNanogallery2 = () => import("./nanogallery2-imports")
+window.initNanogallery2 = () => {
+  if (nanogallery2Loaded) return
+  import("./nanogallery2-imports")
```
Comment (ChristopherChudzicki) [inline course-v2/assets/course-v2.ts:49]: @ibrahimjaved12 By the way: I expect that this import was not actually problematic, but it doesn't hurt to make it happen only once.

Suppose:
```ts
//somefile.js
console.log("Hello from somefile.js")

// main file:

import("./somefile").then(() => {
    console.log('Somefile has loaded, A')
})

import("./somefile").then(() => {
    console.log('Somefile has loaded, B')
})
```
The logs should be:
```

```
Hello from somefile.js
Somefile has loaded, A // The import promise has resolved, so the `then` runes
Somefile has loaded, B // The import promise has resolved (but does not re-run the imported code)
```

When you import a file, its contents only get run once.
